File: tape/sdk/typescript/src/obs.ts
// Observability — structured-log helper + OTel span-name constants.
// Mirrors `tape.obs` in Python and `tape/sdk/go/obs.go`.

export const SPAN_BEGIN_RUN = 'tape.begin_run';
export const SPAN_RESUME_RUN = 'tape.resume_run';
export const SPAN_RECORD_DECISION = 'tape.record_decision';
export const SPAN_BEGIN_EFFECT = 'tape.begin_effect';
export const SPAN_COMPLETE_EFFECT = 'tape.complete_effect';
export const SPAN_RECONCILE_EFFECT = 'tape.reconcile_effect';
export const SPAN_DISPATCH_EFFECT = 'tape.dispatch_effect';
export const SPAN_COMPENSATE = 'tape.compensate';
export const SPAN_REDRIVE = 'tape.redrive';
export const SPAN_AWAIT_SIGNAL = 'tape.await_signal';
export const SPAN_SEND_SIGNAL = 'tape.send_signal';

export const ALL_SPANS: readonly string[] = Object.freeze([
  SPAN_BEGIN_RUN, SPAN_RESUME_RUN, SPAN_RECORD_DECISION,
  SPAN_BEGIN_EFFECT, SPAN_COMPLETE_EFFECT,
  SPAN_RECONCILE_EFFECT, SPAN_DISPATCH_EFFECT,
  SPAN_COMPENSATE, SPAN_REDRIVE,
  SPAN_AWAIT_SIGNAL, SPAN_SEND_SIGNAL,
]);

export const STRUCTURED_FIELDS: readonly string[] = Object.freeze([
  'ts', 'level', 'msg',
  'tenant_id', 'app_name', 'run_id', 'invocation_id', 'session_id',
  'seq', 'effect_key', 'decision_index', 'reactor', 'lease_owner',
]);

/** Emit one structured JSON line to stderr in canonical field order. */
export function logJson(msg: string, fields: Record<string, unknown> = {}, level: string = 'INFO'): void {
  const rec: Record<string, unknown> = { ts: Date.now() / 1000, level, msg };
  for (const [k, v] of Object.entries(fields)) {
    if (v === undefined || v === null || v === '') continue;
    rec[k] = v;
  }
  const ordered: Record<string, unknown> = {};
  for (const k of STRUCTURED_FIELDS) if (k in rec) { ordered[k] = rec[k]; delete rec[k]; }
  for (const [k, v] of Object.entries(rec)) ordered[k] = v;
  process.stderr.write(JSON.stringify(ordered) + '\n');
}

export type SpanEnd = (err?: Error) => void;
export type SpanHook = (name: string, attrs: Record<string, unknown>) => SpanEnd;

let _hook: SpanHook | null = null;

/** Install a span hook. Tracing adapters (e.g. opentelemetry) set this. */
export function setSpanHook(h: SpanHook | null): void { _hook = h; }

/** Open a span via the installed hook (no-op if none). */
export function span(name: string, attrs: Record<string, unknown> = {}): SpanEnd {
  if (!_hook) return () => undefined;
  return _hook(name, attrs);
}

File: tape/sdk/go/obs.go
package tape

// Observability — structured-log helpers and OpenTelemetry span-name
// constants. Mirrors `tape/sdk/python/tape/obs.py`.

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"
	"time"
)

// Span-name constants — use these instead of stringly-typed magic.
const (
	SpanBeginRun         = "tape.begin_run"
	SpanResumeRun        = "tape.resume_run"
	SpanRecordDecision   = "tape.record_decision"
	SpanBeginEffect      = "tape.begin_effect"
	SpanCompleteEffect   = "tape.complete_effect"
	SpanReconcileEffect  = "tape.reconcile_effect"
	SpanDispatchEffect   = "tape.dispatch_effect"
	SpanCompensate       = "tape.compensate"
	SpanRedrive          = "tape.redrive"
	SpanAwaitSignal      = "tape.await_signal"
	SpanSendSignal       = "tape.send_signal"
)

// AllSpans — the canonical span-name list.
var AllSpans = []string{
	SpanBeginRun, SpanResumeRun, SpanRecordDecision,
	SpanBeginEffect, SpanCompleteEffect,
	SpanReconcileEffect, SpanDispatchEffect,
	SpanCompensate, SpanRedrive,
	SpanAwaitSignal, SpanSendSignal,
}

// StructuredFields — the canonical field order for structured log records.
// `LogJSON` emits keys in this order when present.
var StructuredFields = []string{
	"ts", "level", "msg",
	"tenant_id", "app_name", "run_id", "invocation_id", "session_id",
	"seq", "effect_key", "decision_index", "reactor", "lease_owner",
}

var logMu sync.Mutex

// LogJSON — emit one JSON line to stderr, ordered by `StructuredFields`.
// Extra fields are appended after the canonical block.
func LogJSON(msg string, fields map[string]any) {
	rec := map[string]any{
		"ts":    float64(time.Now().UnixNano()) / 1e9,
		"level": "INFO",
		"msg":   msg,
	}
	for k, v := range fields {
		if v == nil || v == "" {
			continue
		}
		rec[k] = v
	}
	// Re-serialise with stable key order.
	ordered := make(map[string]any, len(rec))
	for _, k := range StructuredFields {
		if v, ok := rec[k]; ok {
			ordered[k] = v
			delete(rec, k)
		}
	}
	for k, v := range rec {
		ordered[k] = v
	}
	enc, _ := json.Marshal(ordered)
	logMu.Lock()
	defer logMu.Unlock()
	fmt.Fprintln(os.Stderr, string(enc))
}

// LogLevel — convenience for log_json at a specific level.
func LogLevel(level, msg string, fields map[string]any) {
	if fields == nil {
		fields = map[string]any{}
	}
	fields["level"] = level
	LogJSON(msg, fields)
}

// SpanHook — set by the host to receive Tape span events. Default is nil
// (no-op). A future opentelemetry-go integration would set this from
// `init`.
type SpanHook func(name string, attrs map[string]any) (end func(err error))

var (
	spanHookMu sync.RWMutex
	spanHook   SpanHook
)

// SetSpanHook — install a span hook. Used by tracing adapters; safe to
// call multiple times (last wins).
func SetSpanHook(h SpanHook) {
	spanHookMu.Lock()
	defer spanHookMu.Unlock()
	spanHook = h
}

// Span — open a span via the installed hook (or a no-op). Returns the
// end function the caller defers.
func Span(name string, attrs map[string]any) func(err error) {
	spanHookMu.RLock()
	h := spanHook
	spanHookMu.RUnlock()
	if h == nil {
		return func(error) {}
	}
	return h(name, attrs)
}
